platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/ReusableBucketPool.java

```diff
@@ -0,0 +1,158 @@
+     * locks are used to wait, if the pool is empty (no available buckets), or to wake
+     * up the waiting thread, when a bucket is released back to the pool.
+     */
+    private final AtomicReferenceArray<Object> locks;
+
+    /**
+     * Index in the array of the next available (non-null) bucket. When a bucket is
+     * requested from the pool, it's served from this index, and the corresponding
+     * array element is set to null.
+     *
+     * <p>This index may point to an empty element in the array. It happens, when the
+     * pool is empty. In this case, the thread that requests a new bucket from will be
+     * blocked, until a bucket is released into the array at the index.
+     */
+    private final AtomicInteger nextIndexToGet = new AtomicInteger(0);
+    /**
+     * Index in the array of the next empty slot. When a bucket is released to the pool,
+     * it's put to the array at this index, and the index is increased by one. If another
+     * thread is waiting for a bucket to be released at the given index, it is notified.
+     */
+    private final AtomicInteger nextIndexToRelease = new AtomicInteger(0);
+
+    /**
+     * Creates a new reusable bucket pool of the default size.
+     *
+     * @param serializer Key serializer used by the buckets in the pool
+     */
+    public ReusableBucketPool(final BucketSerializer<K> serializer) {
+        this(DEFAULT_POOL_SIZE, serializer);
+    }
+
+    /**
+     * Creates a new reusable bucket pool of the specified size.
+     *
+     * @param serializer Key serializer used by the buckets in the pool
+     */
+    public ReusableBucketPool(final int size, final BucketSerializer<K> serializer) {
+        poolSize = size;
+        buckets = new AtomicReferenceArray<>(poolSize);
+        locks = new AtomicReferenceArray<>(poolSize);
+        for (int i = 0; i < poolSize; i++) {
+            buckets.set(i, new Bucket<>(serializer.getKeySerializer(), this));
+            locks.set(i, new Object());
+        }
+    }
+
+    /**
+     * Gets a bucket from the pool. If the pool is empty, the calling thread waits
+     * until a bucket is released to the pool.
+     *
+     * @return A bucket that can be used for reads / writes until it's released back
+     * to the pool
+     */
+    public Bucket<K> getBucket() {
+        // Every call to this method is bound to a particular index in the pool
+        final int index = nextIndexToGet.getAndUpdate(t -> (t + 1) % poolSize);
+        // Try optimistic get first
+        Bucket<K> bucket = buckets.getAndSet(index, null);
+        if (bucket == null) {
+            final Object lock = locks.get(index);
+            synchronized (lock) {
+                bucket = buckets.getAndSet(index, null);
+                // Wait until a bucket at the given index is released
+                while (bucket == null) {
+                    try {
+                        lock.wait();
```

Comment:
using Object.wait is discouraged.  Using a Condition for waiting on other threads is strongly encouraged for Java >= 1.5, as it prevents some possible deadlock situations.

Reply:
I tried to use locks instead of monitors here, and synchronization overhead was horrible. It was before I switched from a single lock to lock per index, though. Now with very few (usually zero or one, very rarely two or more) threads waiting on a single index, it may worth trying again. Let me check

Reply:
OK, with a lock per index, it looks good. Let me change Objects to Locks, then.

Reply:
Locks should always outperform Object monitors. If they aren't, then something odd is going on, as the Object monitor (and synchronized keyword) is (at least officially) using the same lock support primitives, just with some extra overhead compared to a ReentrantLock and slightly more opportunity for unexpected deadlock conditions.